build/rewrite.go

```diff
@@ -54,17 +54,55 @@ func allowedSort(name string) bool {
 	return false
 }
 
-// Rewrite applies the high-level Buildifier rewrites to f, modifying it in place.
+// Rewrites controls the rewrites to be applied.
+//
+// If non-nil, the rewrites with the specified names will be run. If
+// nil, a default set of rewrites will be used that is determined by
+// the type (BUILD vs default starlark) of the file being rewritten.
+type Rewriter struct {
+	RewriteSet                      *map[string]struct{}
+	IsLabelArg                      map[string]bool
+	LabelDenyList                   map[string]bool
+	IsSortableListArg               map[string]bool
+	SortableDenylist                map[string]bool
+	SortableAllowlist               map[string]bool
+	NamePriority                    map[string]int
+	StripLabelLeadingSlashes        bool
+	ShortenAbsoluteLabelsToRelative bool
+}
+
 func Rewrite(f *File) {
+	var rewriter = &Rewriter{
+		IsLabelArg:                      tables.IsLabelArg,
+		LabelDenyList:                   tables.LabelDenylist,
+		IsSortableListArg:               tables.IsSortableListArg,
+		SortableDenylist:                tables.SortableDenylist,
+		SortableAllowlist:               tables.SortableAllowlist,
+		NamePriority:                    tables.NamePriority,
+		StripLabelLeadingSlashes:        tables.StripLabelLeadingSlashes,
+		ShortenAbsoluteLabelsToRelative: tables.ShortenAbsoluteLabelsToRelative,
+	}
+	rewriter.Rewrite(f)
+}
+
+func (w *Rewriter) Rewrite(f *File) {
 	for _, r := range rewrites {
-		if !disabled(r.name) {
-			if f.Type&r.scope != 0 {
-				r.fn(f)
-			}
+		// f.Type&r.scope is a bitwise comparison. Because starlark files result in a scope that will
```

Comment:
This is going to result in incorrect behavior where if a file type matches the rewrite scope then the rewrite gets applied even if the rewriter specifies the rewrite set and doesn't include the rewrite.

It should be something like:
```
if disabled(r.name) {
  continue
}
if w.RewriteSet != nil {
  if !rewriteSetContains(w, r.name) {
    continue
  }
} else {
  if f.Type&r.scope == 0 {
    continue
  }
}
r.fn(f, w)
```

Reply:
Good catch! I will add another check on the left side stating that w.Rewrites is empty

Reply:
You shouldn't check for empty rewrites. If the rewrites is non-nil but empty, I would say that should just not perform any rewrites, which is what will happen with my suggested code.

Reply:
nm, I see what you meant